### PR TITLE
Implement LinkedHashMap for BlogMessages

### PR DESCRIPTION
**Pull request by Andrewmob2703-zz**

### CODE REVIEW

**Comment by Andrewmob2703-zz** [comment]
Ignore first implementation of LinkedHashMap commited on July 1

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java:28]: Nit: "map" is okay in this scenario since it should be the only hashmap in BlogHashMap for now. Going forward, it would be good to avoid variable names such as list, map, array, etc. because they are too general, and actual reader will know their they are "map" from their type already.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Consider replacing "isTagSupported = true;" to "return true;" to return early and convey a more clear message to reader: "as long as one tag is matched, it is supported." Also, that way, you don't need to define the "isTagSupported" local variable because you can simply return false when reaching the end of the function.

Comment (peizhaoou) [reply portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Also, when you start to support more tags (which means the array becomes larger), you could consider to use a HashMap for the INTERNAL_TAGS array to check existence/matching in O(1) vs. O(n). You don't have to do it in this PR, but keep that in mind or add a TODO comment for next change, etc.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java:46]: Consider checking error case first and "continue" early to avoid another level of nesting for line 45-54.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Consider using one style of comment for the same method, pick one of /**/ or //.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Consider calling the HashMap's values() method to iterate the values directly instead of using keySet() and get()

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Nit: put single-line statement within {} into a new line.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Nit: no need for the extra parenthesis => map.get(key).size()

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Nit: the comment is kind of confusing. Particularly, what is n0.of.messages?

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Nit: please do not break a "} else {" with comments because it's easy to misread. If the comments in line 98-99 are to describe the else branch, consider:

} else {
  // Otherwise the ...
  // ...
}

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Nit: it would be good to give more details about the argument "limit". I am confused here because I am not entirely sure if "limit" here is to limit the size of the returned LinkedList? E.g., if limit=10, does it mean that the returned list will have at most 10 elements? Please be explicit in the comment.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: "startPoint" is a local variable that is used only once right next line. Would be good to just pass "tagMassageSize - limit" to listIterator() next line.

Comment (peizhaoou) [reply portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Same applies else where.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (martiKallas) [reply portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Just to add to this, it seems that // comments are preferred over multi-line comments in Google code.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (martiKallas) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java:114]: You only really need this function. If tags is empty or null, get all tags. If limit <= 0 don't use a limit. Just document that in the comments of the function. This way you reuse a lot of code and you limit the number of functions that need to be tested.

This also reduces the PR size to a more reasonable level.

File: portfolio/src/main/java/com/google/sps/data/BlogMessage.java
Comment (martiKallas) [inline portfolio/src/main/java/com/google/sps/data/BlogMessage.java]: I would keep this in its own file since it is used in other parts of the code as well, and the uses of BlogMessage aren't dependent on your hashtable API. For example, storing a BlogMessage in Datastore, doesn't also require that Datastore knows about the getMessage function.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (martiKallas) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Normally, linking this into the other parts of the code would be done in a separate PR. For this case it would be for two reasons: to make the PR smaller and more focused on a single goal (implementation), and because this code isn't tested yet. This would be a perfect chance to practice branching a new change off of your current change and migrating this code to that new branch.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (martiKallas) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: You might consider making these tags a public constant in another file if they are going to be used in a number of places.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (Andrewmob2703-zz) [reply portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Got it

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (Andrewmob2703-zz) [reply portfolio/src/main/java/com/google/sps/data/BlogHashMap.java:46]: sounds good

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (Andrewmob2703-zz) [reply portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Got it

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (Andrewmob2703-zz) [reply portfolio/src/main/java/com/google/sps/data/BlogHashMap.java:114]: Sounds good!

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (Andrewmob2703-zz) [reply portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: couldn't find it

Comment (Andrewmob2703-zz) [reply portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Oh just found it in the doc. It's the last method in the doc so my eyes missed it lol

File: .gitignore
Comment (martiKallas) [inline .gitignore]: In cases like this, it is better to create tiny PR's. .gitignore doesn't have anything to do with this PR and a tiny change can be reviewed very quickly.

**Review by martiKallas — CHANGES_REQUESTED** [review]
Make sure you get a review from either Steven or Tayyaba!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (smarquez-jpg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:59]: May not be a problem but I know that when I got the number of comments to display in the doGet instead of the doPost, I ran into the problem where the variable would not update. Other than that LGTM

**Review by smarquez-jpg — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: Nit: use the recommended style guide for consistency:

if (limit <= 0) {
  break;
}

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java:116]: Awesome, good to see you use a constant this way in this case!

File: portfolio/src/main/java/com/google/sps/data/InternalTags.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/InternalTags.java]: Any reason we need to make the methods like setTags() & tagsIsSupported() final?

File: portfolio/src/main/java/com/google/sps/data/InternalTags.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/InternalTags.java]: This could be a static method because they are not specific to any InternalTags object, but a shared method so that one can call "InternalTags.defaultTag()" without instantiating an InternalTags object.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (peizhaoou) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java:38]: Would checking the opposite make the nesting better, e.g.:

if (!InternalTags.tagIsSupported(message)) {
  continue;
}
// Process the case with supported tag with 1 fewer level of nesting.

**Comment by peizhaoou** [comment]
LGTM overall. Please make sure that the current BlogHashMap pass the tests Steven added before you merge.

File: .gitignore
Comment (peizhaoou) [reply .gitignore]: Echo Marti's point. I am fine with leaving it as is for this PR. But going forward, if reviewers make a suggestion like this and there's no strong reason for not doing so, the right way to really address the comment is to revert the change, break it up, and send it for review in a separate PR.

Comment (peizhaoou) [reply .gitignore]: Keep in mind that having smaller and more focused PRs is really a good practice of separation of concerns. It would make your and reviewer's life much easier with focused changes because you want your reviewers to be very focused, and you don't want to deal with too many comments all at once.

**Review by peizhaoou — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (martiKallas) [reply portfolio/src/main/java/com/google/sps/data/BlogHashMap.java:114]: I like how you organized this with the private getMessages functions serving internal purposes!

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (martiKallas) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java:115]: I would recommend against this. AFAIK, it isn't a common convention that if limit <= 0, then choose some other number in it's place. It is slightly more common that if limit == 0 then use no limit. 

Further, DEFAULT_LIMIT doesn't mean much from the user's point of view. I think in this case it is better to imagine you are writing a library that will be used by someone other than your pod. Then ask if they would want a default value and why.

Always try to make sure default values are meaningful rather than arbitrary. If you can't come up with a meaningful default value, then it is usually better to handle that situation in another way like throwing an error, or re-defining the function.

File: portfolio/src/main/java/com/google/sps/data/BlogHashMap.java
Comment (martiKallas) [inline portfolio/src/main/java/com/google/sps/data/BlogHashMap.java]: nit: At least for C++, the common style is to bracket a parameter name with vertical slashes: If |tags| are not specified...

**Review by martiKallas — APPROVED** [review]
I think this is really nice work! I just left a couple of minor comments.

I think you can do a little better with splitting things up into smaller PRs as well. Each file you changed could have been it's own PR, although combining InternalTags and BlogHashMap makes sense.

**Pull request merged**
